worker/changestream/eventmultiplexer/eventmultiplexer.go

```diff
@@ -0,0 +1,296 @@
+				continue
+			}
+
+			// Register filters to route changes matching the subscription criteria to
+			// the newly created subscription.
+			for _, opt := range subOpt.opts {
+				namespace := opt.Namespace()
+				e.subscriptionsByNS[namespace] = append(e.subscriptionsByNS[namespace], &eventFilter{
+					subscriptionID: sub.id,
+					changeMask:     opt.ChangeMask(),
+					filter:         opt.Filter(),
+				})
+				sub.topics[namespace] = struct{}{}
+			}
+
+		case subscriptionID := <-e.unsubscriptionCh:
+			sub, found := e.subscriptions[subscriptionID]
+			if !found {
+				continue
+			}
+
+			for topic := range sub.topics {
+				var updatedFilters []*eventFilter
+				for _, filter := range e.subscriptionsByNS[topic] {
+					if filter.subscriptionID == subscriptionID {
+						continue
+					}
+					updatedFilters = append(updatedFilters, filter)
+				}
+				e.subscriptionsByNS[topic] = updatedFilters
+			}
+
+			delete(e.subscriptions, subscriptionID)
+			delete(e.subscriptionsAll, subscriptionID)
+
+			sub.close()
+		}
+	}
+}
+
+func (e *EventMultiplexer) gatherSubscriptions(ch changestream.ChangeEvent) []*subscription {
+	subs := make(map[uint64]*subscription)
+
+	for id := range e.subscriptionsAll {
+		subs[id] = e.subscriptions[id]
+	}
+
+	traceEnabled := e.logger.IsTraceEnabled()
+	for _, subOpt := range e.subscriptionsByNS[ch.Namespace()] {
+		if _, ok := subs[subOpt.subscriptionID]; ok {
+			continue
+		}
+
+		if (ch.Type() & subOpt.changeMask) == 0 {
+			continue
+		}
+
+		if !subOpt.filter(ch) {
+			if traceEnabled {
+				e.logger.Tracef("filtering out change: %v", ch)
+			}
+			continue
+		}
+
+		if traceEnabled {
+			e.logger.Tracef("dispatching change: %v", ch)
+		}
+
+		subs[subOpt.subscriptionID] = e.subscriptions[subOpt.subscriptionID]
+	}
+
+	// By collecting the subs within a map to ensure that a sub can be only
+	// called once, we actually gain random ordering. This prevents subscribers
+	// from depending on the order of dispatches.
+	results := make([]*subscription, 0, len(subs))
+	for _, sub := range subs {
+		results = append(results, sub)
+	}
+	return results
+}
+
+// dispatchSet fans out the subscription requests against a given term of changes.
+// Each subscription signals the change in a asynchronous fashion, allowing
+// a subscription to not block another change within a given term.
+func (e *EventMultiplexer) dispatchSet(changeSet map[*subscription]ChangeSet) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+
+	grp, grpCtx := errgroup.WithContext(e.catacomb.Context(ctx))
+	grp.SetLimit(len(changeSet))
+
+	for sub, changes := range changeSet {
+		sub, changes := sub, changes
+
+		grp.Go(func() error {
+			// Pass the context of the catacomb with the deadline to the
+			// subscription. This allows the subscription to be cancelled
+			// if the catacomb is dying or if the deadline is reached.
+			return sub.signal(changes, grpCtx.Done(), ctx.Done())
```

Comment:
grpCtx is a child of ctx, so we don't need to pass both

Reply:
infact I think what is most correct here:
- move the context.WithTimeout into the sub.signal method
- pass the grpCtx rather than the done channel to sub.signal
- each signal command then waits on a single ctx
- if the errors.Is(ctx.Err(), context.DeadlineExceeded) then we should unsubscribe that subscription.

Reply:
I think that's probably a better design.